forkmanager/fork_manager.go

```diff
@@ -0,0 +1,233 @@
+package forkmanager
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+)
+
+/*
+Regarding whether it is okay to use the Singleton pattern in Go, it's a topic of some debate.
+The Singleton pattern can introduce global state and make testing and concurrent programming more challenging.
+It can also make code tightly coupled and harder to maintain.
+In general, it's recommended to favor dependency injection and explicit collaboration over singletons.
+
+However, there might be scenarios where the Singleton pattern is still useful,
+such as managing shared resources or maintaining a global configuration.
+Just be mindful of the potential drawbacks and consider alternative patterns when appropriate.
+*/
+
+var (
+	forkManagerInstance     *forkManager
+	forkManagerInstanceLock sync.Mutex
+)
+
+type forkManager struct {
+	lock sync.Mutex
+
+	forkMap     map[ForkName]*Fork
+	handlersMap map[ForkHandlerName][]ForkActiveHandler
```

Comment:
Shouldn't this be a part of a fork? Why do we need all the handlers inside of the `ForkManager`? I assumed that all of the handlers will be assigned to a specific fork.

Reply:
its optimization. binary search is used to determine which handler should be used.
For example there are 3 forks:
(Fork=a, StartingBlock=0)
(Fork=b, StartingBlock=10)
(Fork=c, StartingBlock=20)

lets say handlers are registered for fork a, b but not for c

If you want to retrieve handler for block < 10, then you should get  handler for fork a
If you want to retrieve handler for block >= 10 (even block lets say 66), then you should get  handler for fork b